_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+2. The Authorization Endpoint responds as usual but records the "code_challenge" and the transformation method.
+3. The Client then sends the authorization code in the Access Token Request as usual but includes the "code_verifier" secret generated in the first step.
+4. The Authorization Server transforms "code_verifier" and compares it to the recorded "code_challenge". Access is denied if they are not equal.
+
+The latest [Security BCP](https://www.ietf.org/archive/id/draft-ietf-oauth-security-topics-24.html) states that PKCE should be enabled for all types of clients, public and confidential (browser-based applications, mobile applications, native applications, and secure server applications).
+
+## Spring Security for Authorization Code Flow
+
+You can experiment with how to configure the authorization code flow configuration by creating a simple Spring Boot web application, following the step-by-step guide in the following sections.
+
+
+### Create a simple Spring Boot application
+
+With Spring Initializr and curl, create Spring Boot project:
+
+```shell
+curl -v "https://start.spring.io/starter.zip?\
+bootVersion=3.2.3&\
+language=java&\
+packaging=jar&\
+javaVersion=17&\
+type=gradle-project&\
+dependencies=web,okta,thymeleaf&\
+groupId=com.example&\
+artifactId=spring-web&\
+packageName=com.example.demo" \
+--output spring-web.zip
+```
+
+Then extract the content:
+
+```shell
+unzip spring-web.zip -d spring-web
+```
+
+If you inspect the contents of `build.gradle`, you will find the Okta Spring Boot Starter dependency is included. The Okta Spring Boot Starter simplifies the process of adding authentication into your Spring Boot application by auto-configuring the necessary classes and adhering to best practices, eliminating the need for you to do it manually. It leverages the OAuth 2.0 and OpenID Connect protocols for user authentication.
+
+Add the following additional dependencies to `build.gradle`:
+
+```groovy
+// build.gradle
+dependencies {
+    ...
+    implementation 'me.paulschwarz:spring-dotenv:4.0.0'
+    ...
+}    
+```
+
+Create the `index.html` template at `src/main/resources/templates` with the following content:
+
+```html
+<!-- src/main/resources/templates/index.html -->
+<!doctype html>
+<html lang="en">
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Bootstrap demo</title>
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-T3c6CoIi6uLrA9TneNEoa7RxnatzjcDSCmG1MXxSR1GAsXEV/Dwwykc2MPK8M2HN" crossorigin="anonymous">
+    <link rel="stylesheet" href="styles.css">
+</head>
+<body>
+
+<th:block th:replace="~{fragments/navigation.html}" />
+
+
+<div class="container my-5">
+    <h1>Hello, world!</h1>
```

Comment:
please remove the bootstrap boilerplate and add something more relevant. please see this for example: https://github.com/oktadev/auth0-spring-boot-passkeys-demo/blob/main/src/main/resources/templates/home.html

Reply:
Updated